mars/services/task/analyzer/fusion.py

```diff
@@ -12,101 +12,165 @@
+                and not chunk.op.scheduling_hint.can_be_fused()
+            )
+        ):
+            return False
+        return True
+
+    def _color_init_nodes(self) -> Dict[OperandType, int]:
+        # for initial op with same band but different priority
+        # we color them w/ different colors,
+        # to prevent from wrong fusion.
+        band_priority_to_colors = dict()
+        for chunk, band in self.chunk_to_bands.items():
+            band_priority = (band, chunk.op.priority)
+            if band_priority not in band_priority_to_colors:
+                band_priority_to_colors[band_priority] = self._next_color()
+
+        band_priority_to_color_list = defaultdict(list)
+        for (band, priority), color in band_priority_to_colors.items():
+            band_priority_to_color_list[band, priority].append(color)
+        color_to_size = defaultdict(lambda: 0)
+        op_to_colors = dict()
+        for chunk, band in self.chunk_to_bands.items():
+            priority = chunk.op.priority
+            color = band_priority_to_color_list[band, priority][-1]
+            size = color_to_size[color]
+            if size >= self.initial_same_color_num:
+                color = self._next_color()
+                band_priority_to_color_list[band, priority].append(color)
+            color_to_size[color] += 1
+            op_to_colors[chunk.op] = color
+        return op_to_colors
+
+    def color(self) -> Dict[ChunkType, int]:
+        chunk_to_colors = dict()
+
+        # step 1: Coloring the initial nodes according to the bands that assigned by assigner
+        op_to_colors = self._color_init_nodes()
+
+        # step2: Propagate color in the topological order,
+        # if the input nodes have same color, color it with the same color;
+        # otherwise, color with a new color.
+        chunk_to_is_broadcaster = dict()
+        for chunk in self.chunk_graph.topological_iter():
+            if self.chunk_graph.count_successors(chunk) > self.successor_same_color_num:
+                # is broadcaster
+                chunk_to_is_broadcaster[chunk] = True
+
+            if chunk.op in op_to_colors:
+                # colored
+                chunk_to_colors[chunk] = op_to_colors[chunk.op]
                 continue
-            selected = [v]
-            # add successors
-            while self._graph.count_predecessors(cur_node) == 1 and not isinstance(
-                cur_node.op, (VirtualOperand, Fetch)
-            ):
-                selected.append(cur_node)
-                if (
-                    self._graph.count_successors(cur_node) != 1
-                    or cur_node in result_chunk_set
-                ):
-                    break
+
+            predecessors = self.chunk_graph.predecessors(chunk)
+            pred_colors = {op_to_colors[pred.op] for pred in predecessors}
+            if len(predecessors) == 1 and chunk_to_is_broadcaster.get(predecessors[0]):
```

Comment:
It is possible that a successor accepts a non-broadcaster input with a broadcast one. Under this scenario it is possible to get unexpected color result. Also need to consider if allocating a new color directly is oversimplified. Is it possible to use some grouping strategy instead or add some TODO items?